### PR TITLE
[VICPOLSL-62] - Access the station API

### PR DESCRIPTION
**Pull request by gargsuchi**
https://salsadigital.atlassian.net/browse/VICPOLSL-62

### CODE REVIEW

File: src/StationAPIConnector.php [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        'name' => 'STATION_LOCATION',
+        'attributes' => [
+          [
+            'name' => 'ACTIVE',
+            'value' => ['Y'],
+            'caseSensitive' => FALSE,
+          ],
+        ],
+        'startIndex' => 1,
+        'recordCount' => 5,
```
Comment (gargsuchi) [inline src/StationAPIConnector.php]: @g2410g Why are we limiting this to 5?

File: src/StationAPIConnector.php [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+      unset($station['attributes']);
+      $final_stations['records'][$key] = $station;
+    }
+
+    $file_save_path_stream_directory = 'public://';
+    $this->fileSystem->prepareDirectory($file_save_path_stream_directory, FileSystemInterface::CREATE_DIRECTORY | FileSystemInterface::MODIFY_PERMISSIONS);
+    $fileLocation = $file_save_path_stream_directory . '/' . 'stations.json';
+    // Save the processed JSON file.
+    $file = file_save_data(json_encode($final_stations), $fileLocation, FileSystemInterface::EXISTS_REPLACE);
+  }
```
Comment (gargsuchi) [inline src/StationAPIConnector.php:302]: @g2410g We had talked about extracting the taxonomy terms etc. using this loop. Where is that code?

Comment (gargsuchi) [reply src/StationAPIConnector.php:302]: Once that is done, we will not need the assets dir. We should change the paths for  state.csv/ accessibility.csv and specialityservicesandfacility.csv to `public://`

File: src/Commands/TideStationLocatorMigrationCommands.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+   * Import State terms.
+   *
+   * @usage drush migrate-state
+   *   Import stations.
+   *
+   * @command migrate-states
+   * @aliases m-states
+   * @validate-module-enabled tide_station_locator
+   */
+  public function importStates() {
```
Comment (gargsuchi) [inline src/Commands/TideStationLocatorMigrationCommands.php]: We dont need this. 
We can easily use `drush migrate:import station_state_terms`

File: src/Commands/TideStationLocatorMigrationCommands.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+   *
+   * @usage drush migrate-speciality
+   *   Import speciality terms.
+   *
+   * @command migrate-speciality
+   * @aliases m-speciality
+   * @validate-module-enabled tide_station_locator
+   */
+  public function importSpeciality() {
+    tide_station_locator_migrate_speciality();
```
Comment (gargsuchi) [inline src/Commands/TideStationLocatorMigrationCommands.php]: We dont need this. Same as above

File: tide_station_locator.module [hunk 5]
```diff
@@ -4,3 +4,90 @@
...
+  catch (Exception $e) {
+    // Something went wrong somewhere.
+    watchdog_exception('tide_station_locator', $e);
+  }
+}
+
+/**
+ * Function to migrate state terms.
+ */
+function tide_station_locator_migrate_state() {
```
Comment (gargsuchi) [inline tide_station_locator.module]: We dont need this.

File: tide_station_locator.module [hunk 6]
```diff
@@ -4,3 +4,90 @@
...
+  catch (Exception $e) {
+    // Something went wrong somewhere.
+    watchdog_exception('tide_station_locator', $e);
+  }
+}
+
+/**
+ * Function to migrate speciality terms.
+ */
+function tide_station_locator_migrate_speciality() {
```
Comment (gargsuchi) [inline tide_station_locator.module]: We dont need this.

File: tide_station_locator.module [hunk 7]
```diff
@@ -4,3 +4,90 @@
...
+  catch (Exception $e) {
+    // Something went wrong somewhere.
+    watchdog_exception('tide_station_locator', $e);
+  }
+}
+
+/**
+ * Function to migrate accessibility terms.
+ */
+function tide_station_locator_migrate_accessibility() {
```
Comment (gargsuchi) [inline tide_station_locator.module]: We dont need this.

**Review by gargsuchi — COMMENTED** [review]
Hi

**Review by gargsuchi — COMMENTED** [review]
Hi @g2410g 
Please check my comments on the PR.

File: src/StationAPIConnector.php [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+    $config = $this->configFactory->getEditable('tide_station_locator.settings');
+
+    if ($delta && !empty($config->get('last_api_access'))) {
+      $payload = [
+        'name' => 'STATION_LOCATION',
+        "productKey" => -1,
+        'attributes' => [
+          [
+            'name' => 'RECORD_MODDATE',
+            'value' => [$config->get('last_api_access')],
```
Comment (gargsuchi) [inline src/StationAPIConnector.php]: Do a config->get once and use here and at #174

File: src/StationAPIConnector.php [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+    return $payload;
+  }
+
+  /**
+   * Helper method to get the 'last_access_api'.
+   *
+   * @return string
+   *   Config value last_api_access.
+   */
+  public function get_last_access_api() {
```
Comment (gargsuchi) [inline src/StationAPIConnector.php]: Where are we using this? why are we setting the value to now if no value is found?

**Comment by vincent-gao** [comment]
hey @gargsuchi / @g2410g 
The failed build was caused by Claro and Ckeditor5 upgrade, the contribute maintainers are currently working on upgrading the modules in favour of Ckeditor5, so, we inevitably rely on patches, composer aliases, and ticket repositories for upgrading Ckeditor5 ourselves. However, during the upgrade to Drupal 10, we will revisit these changes, removing them if possible.

**Comment by vincent-gao** [comment]
failed again, because of the lint

File: src/StationAPIConnector.php [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+
+      return [
+        'clientId' => $client,
+        'accessToken' => $accessToken,
+        'rdm_auth' => $rdm_auth,
+      ];
+    }
+    catch (BadResponseException $e) {
+      $this->logger->get('tide_station_locator connect API.')
+        ->error('BadResponse error: @error', ['@error' => $e->getResponse()]);
```
Comment (gargsuchi) [inline src/StationAPIConnector.php]: @g2410g If there is an error in the call, we are getting this error:
`Error: Object of class GuzzleHttp\Psr7\Response could not be converted to string in`

**Pull request merged**